### PR TITLE
task osco-to-oscal

### LINKED ISSUE #260: Transform OSCO yaml into NIST OSCAL-like Assessment Results Observations list json as a trestle task
## Issue description / feature objectives

Assumptions:

- input directory contains a collection of OSCO produced yaml files
- input directory contains, optionally, a metadata file comprising additional reference data named _oscal-metadata.yaml_
- output directory exists or is creatable

Approach:

- create the output directory, if necessary
- for each input OSCO yaml file:
    - discover the rule and result pair entries contained within
    - construct an observation for each entry, including any additional related information from the metadata file if present
    - create the corresponding output file, only if the file does not already exist or overwrite=true (the default)

Example CLI:
`trestle task osco-to-oscal -c /home/degenaro/git/degenaro.evidence-locker/osco/demo-osco-to-oscal.config`

Example config:
```
[task.osco-to-oscal]

input-dir =  /home/degenaro/git/degenaro.evidence-locker/osco/input
output-dir = /home/degenaro/git/degenaro.evidence-locker/osco/output
output-overwrite = true
```

### PR DESCRIPTION
**Pull request by degenaro**
## Types of changes

- \[ \] Bug fix (non-breaking change which fixes an issue)
- \[x\] New feature (non-breaking change which adds functionality)
- \[ \] Breaking change (fix or feature that would cause existing functionality to change)
- \[x\] My code follows the code style of this project.
- \[ \] My change requires a change to the documentation.
- \[ \] I have updated the documentation accordingly.
- \[x\] I have added tests to cover my changes.
- \[x\] All new and existing tests passed.
- \[x\] All commits are signed-off.

## Description

- Provide transformation function from dict of OSCO YAML (XCCDF) to dict of OSCAL-like JSON (observations list).
- Accept transformation function metadata YAML to enhance produced OSCAL-like JSON.
- Provide trestle task osco-to-oscal to read input OSCO files from input directory and write OSCAL-like transformation data to corresponding files in output directory.
- Provide test-cases covering transformation function and task.

## Comments

The OpenShift Compliance Operator (see https://docs.openshift.com/container-platform/4.6/security/compliance_operator/compliance-operator-understanding.html) employs OpenSCAP, a NIST-certified tool, to scan and enforce security policies. The scan results are available as YAML. The osco-to-oscal function and task perform a transformation from this YAML format to NIST OSCAL-like JSON as a list of observations, each observation conforming to the observation component of Assessment Results (see https://pages.nist.gov/OSCAL/documentation/schema/assessment-results-layer/assessment-results/json-schema/#oscal-ar-json_observation).

closes #260

### CODE REVIEW

File: trestle/utils/osco.py [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+            results = resource['data']['results']
+            if results.startswith('<?xml'):
+                pass
+            else:
+                results = bz2.decompress(base64.b64decode(results))
+            logger.debug('========== <results> ==========')
+            logger.debug(results)
+            logger.debug('========== </results> ==========')
+            self._parse_xml(results, self._metadata['name'])
+            break
```
Comment (butler54) [inline trestle/utils/osco.py:241]: while True is an awkward construct here for something that is non-looping. You can just
```python:
if bad_condition:
     return
``` 
instead.

File: tests/trestle/tasks/osco_test.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+import yaml
+
+stem_in = pathlib.Path('tests/data/tasks/osco/input')
+stem_in_no_metadata = pathlib.Path('tests/data/tasks/osco/input-no-metadata')
+stem_out = pathlib.Path('tests/data/tasks/osco/output')
+
+mock_uuid4_value = uuid.UUID('d4cf7a88-cea7-4667-9924-8d278dc843df')
+
+
+def test_function_osco_get_observations(tmpdir):
```
Comment (butler54) [inline tests/trestle/tasks/osco_test.py:35]: if you replace tmpdir with tmp_path it is the same thing but a pathlib object - which will remove the need for line 45.

File: tests/trestle/tasks/osco_test.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    """Test OSCO to OSCAL transformation."""
+    idata = _load_yaml(stem_in, 'ssg-ocp4-ds-cis-111.222.333.444-pod.yaml')
+    metadata = _load_yaml(stem_in, 'oscal-metadata.yaml')
+    with patch('uuid.uuid4') as mock_uuid4:
+        mock_uuid4.return_value = mock_uuid4_value
+        observations, analysis = osco.get_observations(idata, metadata)
+    expected = _load_json(stem_out, 'osco-pod-oscal.json')
+    tfile = tmpdir / 'osco-pod-oscal.json'
+    import pathlib
```
Comment (butler54) [inline tests/trestle/tasks/osco_test.py:44]: You have already imported pathlib in line 20.

**Review by butler54 — CHANGES_REQUESTED** [review]
So a few structural elements to think through with respect to this PR.
1. We (I) have not thought so far with respect to code for specific tasks. The thought here is w.r.t. osco.py and the partial asessment results structure that you have. My gut feel is we should probably organise this info together? (I think into the trestle.tasks directory would be fine). The reason being: That most users will not know what to do with a partial assessment results structure is you can probably refactor into one of your modules (osco.py) to simplify things.
2. the typing in places is a little wacky.
3. have a look at the google style docs for doing function / class variable annotations (not a big deal).
4. The one question I had with the PartialAssessmentResults class is that you are going to get on disk a structure like this:
```json
{"partial-assessment-results":{"observations" : [ ]}}
```
If in an ideal world what we really want is just
```json
{"observations" : [ ]}}
```
Let me know and we can create a helper method to deal with this.

File: trestle/utils/osco.py (hunk 1 above)
Comment (degenaro) [reply trestle/utils/osco.py:241]: I dislike more than one exit point, but I understand the awkwardness and will comply.

File: trestle/utils/osco.py [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+from trestle.oscal.assessment_results import Observation
+from trestle.oscal.assessment_results import Property
+from trestle.oscal.assessment_results import RelevantEvidence
+from trestle.oscal.assessment_results import Subject
+from trestle.oscal.assessment_results_partial import AssessmentResultsPartial
+
+logger = logging.getLogger(__name__)
+
+
+def get_observations(idata: Dict, oscal_metadata: Dict) -> (AssessmentResultsPartial, Dict):
```
[the remark this answers is not in the document]
Comment (degenaro) [reply trestle/utils/osco.py]: Fixed.

File: trestle/utils/osco.py [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+    Optional ```oscal_metadata``` helps more completely formulate each observation.
+
+    Returned ```odata``` comprises list of OSCAL-like Assessment Results observations.
+    Returned ```analysis``` comprises dict of statistics.
+    """
+    rules = Rules(idata)
+    return _get_observations(rules, oscal_metadata)
+
+
+def _get_observations(rules: Dict, oscal_metadata: Dict) -> (AssessmentResultsPartial, Dict):
```
[the remark this answers is not in the document]
Comment (degenaro) [reply trestle/utils/osco.py]: Fixed.

File: trestle/utils/osco.py [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+    if classification is not None:
+        external_data['class'] = classification
+    external_data['name'] = name
+    external_data['value'] = value
+    value = Property(**external_data)
+    logger.debug(f'_get_property: {value}')
+    return value
+
+
+def _get_entry(oscal_metadata: Dict, name: str) -> Union[Dict, None]:
```
[the remark this answers is not in the document]
Comment (degenaro) [reply trestle/utils/osco.py]: Fixed.

File: trestle/utils/osco.py [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+        if name in oscal_metadata:
+            value = oscal_metadata[name]
+    logger.debug(f'_get_entry: {value}')
+    return value
+
+
+class Rules():
+    """Create and accumulate list of rule + result pairs with associated metadata."""
+
+    def __init__(self, content: Dict) -> Any:
```
[the remark this answers is not in the document]
Comment (degenaro) [reply trestle/utils/osco.py]: Fixed.

File: trestle/oscal/assessment_results_partial.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# generated by datamodel-codegen:
+#   filename:  oscal_assessment-results_schema.json
+
+from __future__ import annotations
```
[the remark this answers is not in the document]
Comment (degenaro) [reply trestle/oscal/assessment_results_partial.py]: Fixed. (file is removed)

File: trestle/utils/osco.py [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+def _get_observations(rules: Dict, oscal_metadata: Dict) -> (AssessmentResultsPartial, Dict):
+    """Produce an OSCAL-like partial results object comprising a collection of Observations."""
+    observation_list = []
+    for rule in rules.instances:
+        observation = _get_observation(rule, oscal_metadata)
+        observation_list.append(observation)
+    external_data = {
+        'observations': observation_list,
+    }
+    value = AssessmentResultsPartial(**external_data)
```
[the remark this answers is not in the document]
Comment (degenaro) [reply trestle/utils/osco.py]: Fixed.

File: trestle/utils/osco.py [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+    external_data = {}
+    external_data['uuid'] = str(uuid.uuid4())
+    external_data['description'] = rule['idref']
+    external_data['title'] = rule['idref']
+    external_data['relevant-evidence'] = _get_relevant_evidence(rule, oscal_metadata)
+    external_data['subjects'] = _get_subjects(rule, oscal_metadata)
+    external_data['methods'] = ['TEST-AUTOMATED']
+    if len(external_data['subjects']) == 0:
+        del external_data['subjects']
+    value = Observation(**external_data)
```
[the remark this answers is not in the document]
Comment (degenaro) [reply trestle/utils/osco.py]: Fixed.

File: trestle/utils/osco.py (hunk 1 above)
Comment (degenaro) [reply trestle/utils/osco.py]: Fixed.

File: tests/trestle/tasks/osco_test.py (hunk 2 above)
Comment (degenaro) [reply tests/trestle/tasks/osco_test.py]: Fixed.

File: tests/trestle/tasks/osco_test.py (hunk 3 above)
Comment (degenaro) [reply tests/trestle/tasks/osco_test.py]: Fixed.

**Comment by degenaro** [comment]
Claim is that all items have been addressed, except for the disk structure issue.

**Review by butler54 — CHANGES_REQUESTED** [review]
See if the open comments appear on your side.

File: trestle/oscal/assessment_results_partial.py [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+
+from typing import List, Optional
+
+from pydantic import Field
+from trestle.core.base_model import OscalBaseModel
+from trestle.oscal.assessment_results import Observation
+
+
+class AssessmentResultsPartial(OscalBaseModel):
+    observations: Optional[List[Observation]] = Field(None, min_items=1)
```
[the remark this answers is not in the document]
Comment (degenaro) [reply trestle/oscal/assessment_results_partial.py]: Fixed. In the manner suggested.

File: trestle/tasks/osco_to_oscal.py [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+    """
+
+    name = 'osco-to-oscal'
+
+    def __init__(self, config_object: Optional[configparser.SectionProxy]) -> None:
+        """
+        Initialize trestle task osco-to-oscal.
+
+        Attributes:
+            config_object: Config section associated with the task.
```
[the remark this answers is not in the document]
Comment (degenaro) [reply trestle/tasks/osco_to_oscal.py:49]: Fixed.

File: trestle/tasks/osco_to_oscal.py [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+            if len(metadata) == 0:
+                logger.debug(f'no metadata: {imeta}.')
+            # examine each file in the input folder
+            for ifile in sorted(ipth.iterdir()):
+                # skip enhancing oscal metadata
+                if ifile.name == imeta:
+                    continue
+                # ignore non-yaml files
+                if ifile.suffix != '.yml':
+                    if ifile.suffix != '.yaml':
```
[the remark this answers is not in the document]
Comment (degenaro) [reply trestle/tasks/osco_to_oscal.py:124]: Fixed.

[the remark this answers is not in the document]
Comment (degenaro) [reply trestle/tasks/osco_to_oscal.py:124]: Fixed.

File: trestle/tasks/osco_to_oscal.py [hunk 14]
```diff
@@ -0,0 +1,252 @@
...
+        observations.oscal_write(ofile)
+    
+    def _calculate_ofile(self, ifn: str, opth: pathlib.Path) -> pathlib.Path:
+        """Synthesize output file path+name."""
+        ofn = ifn
+        ofn = ofn.rsplit('.yaml')[0]
+        ofn = ofn.rsplit('.yml')[0]
+        ofn += '-oscal.json'
+        ofile = opth / ofn
+        return ofile
```
[the remark this answers is not in the document]
Comment (degenaro) [reply trestle/tasks/osco_to_oscal.py:240]: Fixed.

File: trestle/tasks/osco_to_oscal.py [hunk 15]
```diff
@@ -0,0 +1,252 @@
...
+                    logger.info(f'result types: {analysis["result_types"]}')
+            return TaskOutcome('success')
+        logger.error(f'config missing')
+        return TaskOutcome('failure')
+    
+    def _read_content(self, ifile: pathlib.Path):
+        """Read the contents of a yaml file."""
+        with open(ifile, 'r+') as fp:
+            data = fp.read()
+            content = yaml.full_load(data)
```
[the remark this answers is not in the document]
Comment (degenaro) [reply trestle/tasks/osco_to_oscal.py:221]: Fixed.

File: trestle/tasks/osco_to_oscal.py [hunk 16]
```diff
@@ -0,0 +1,252 @@
...
+        return ofile
+    
+    def _get_metadata(self, mfile: pathlib.Path, default_metadata: Dict) -> Dict:
+        """Get metadata, if it exists."""
+        metadata = default_metadata
+        try:
+            with open(mfile, "r") as fp:
+                metadata = yaml.full_load(fp)
+        except:
+            logger.debug(traceback.format_exc())
```
[the remark this answers is not in the document]
Comment (degenaro) [reply trestle/tasks/osco_to_oscal.py:249]: Fixed.

[the remark this answers is not in the document]
Comment (degenaro) [reply trestle/tasks/osco_to_oscal.py:249]: The possibilities here are: 1) a metadata file name was specified 2) no metadata file name was specified, therefore use the default name. In either case, if the metadata file is not usable (cannot read it, does not exist, not yaml, etc.) we simply do not use it. You will see in the output that your Observations are not augmented. If you want to know why, run with the debug option and you'll see the stack trace. 

I'm open to suggestions on how to improve.

File: trestle/utils/osco.py [hunk 17]
```diff
@@ -0,0 +1,314 @@
...
+logger = logging.getLogger(__name__)
+
+
+def get_observations(idata: Dict, oscal_metadata: Dict) -> (AssessmentResultsPartial, Dict):
+    """
+    Transform OSCO yaml to NIST OSCAL json with statistics.
+
+    Required ```idata``` is dict representation of OSCO yaml.
+    Optional ```oscal_metadata``` helps more completely formulate each observation.
+
```
[the remark this answers is not in the document]
Comment (degenaro) [reply trestle/utils/osco.py]: Fixed.

File: trestle/tasks/osco_to_oscal.py [hunk 18]
```diff
@@ -0,0 +1,252 @@
...
+        logger.info('      type: <type>')
+        logger.info('      title: <title>')
+        logger.info('      properties: ')
+        logger.info('        target: <target>')
+        logger.info('        cluster-name: <cluster-name>')
+        logger.info('        cluster-type: <cluster-type>')
+        logger.info('        cluster-region: <cluster-region>')
+        logger.info('')
+        logger.info('Augmentation occurs when the name specified in the metadata entry of the osco .yaml file matches an entry <name> specified in the input-metadata .yaml file, if any.')
+        logger.info('All entries in the input-metadata .yaml are optional. For example, locker and its value can be omitted.')
```
[the remark this answers is not in the document]
Comment (degenaro) [reply trestle/tasks/osco_to_oscal.py:89]: Fixed. By adding the following:

Likewise, any or all of cluster-name, -type, and -region may be omitted if not applicable.

**Comment by degenaro** [comment]
Claim is that all issues above are now addressed. Please review, thx!

**Review by butler54 — APPROVED** [review]
LGTM

**Pull request merged**